sources/ghostdog/src/main.rs

```diff
@@ -319,6 +379,302 @@ lazy_static! {
+                    }
+                }
+                "Link layer" => {
+                    if let Some(port) = &mut current_port {
+                        port.link_layer = value.to_string();
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // After all the line parsing, put the last port into the Vec.
+        if let Some(port) = current_port {
+            ca.ports.push(port);
+        }
+        Ok(ca)
+    }
+}
+
+/// Finds the device name using the nvme-cli
+fn get_ibstat_for_device(device: OsString) -> Result<InfinibandDevice> {
+    // let device_name = device.to_string_lossy();
+    let device_name = device.into_string().unwrap();
+    // ibstat can take a device as the first arg to limit output
+    let output = Command::new(IBSTAT_CLI_PATH)
+        .args([device_name.clone()])
+        .output()
+        .context(error::IbstatCommandSnafu {
+            device: device_name,
+        })?;
+
+    parse_ibstat_output(&output.stdout)
+}
+
+fn parse_ibstat_output(input: &[u8]) -> Result<InfinibandDevice> {
+    let ibstat_str = str::from_utf8(input).context(error::ReadIbstatOutputSnafu)?;
+    let ib_device =
+        InfinibandDevice::from_str(ibstat_str).context(error::ParseIbstatOutputSnafu)?;
+    Ok(ib_device)
+}
+
+/// write_guid will read the two files that need to be coordinated, search them for
+/// existing references, and replace the reference if found. If the reference doesn't
+/// exist, it will append to the end of the file the correct reference. This uses a
+/// tempfile to avoid reading and writing to the same file and copies the temp file
+/// over the exiting file when complete.
+fn write_guid(guid: String) -> Result<()> {
+    let mut found_fabric_manager_port = false;
+    let mut found_nvlsm_port = false;
+
+    // Search and update fabricmanager.cfg
+    let mut fm_config = NamedTempFile::new().context(error::CreateTemporaryFileSnafu)?;
+
+    if let Ok(lines) = read_lines(FABRIC_MANAGER_CONFIG_PATH) {
+        for line in lines.map_while(std::result::Result::ok) {
+            if line.contains("FM_SM_MGMT_PORT_GUID") {
+                found_fabric_manager_port = true;
+                writeln!(fm_config, "FM_SM_MGMT_PORT_GUID={}", guid.clone()).context(
+                    error::WriteFileSnafu {
+                        path: FABRIC_MANAGER_CONFIG_PATH.to_string(),
+                    },
+                )?;
+            } else {
+                writeln!(fm_config, "{}", line).context(error::WriteFileSnafu {
+                    path: FABRIC_MANAGER_CONFIG_PATH.to_string(),
+                })?;
+            }
+        }
+    }
+    // If there was no reference replaced, append to the end of the file
+    if !found_fabric_manager_port {
+        writeln!(fm_config, "FM_SM_MGMT_PORT_GUID={}", guid.clone()).context(
+            error::WriteFileSnafu {
+                path: FABRIC_MANAGER_CONFIG_PATH.to_string(),
+            },
```

Comment:
This is confusing since you're not actually writing to this file, but a temporary one.